### PR TITLE
[reland] Perform appropriate CUDA stream synchronization in distributed autograd.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#53929 [reland] Perform appropriate CUDA stream synchronization in distributed autograd.**

The local autograd engine performs appropriate stream synchronization
between autograd nodes in the graph to ensure a consumer's stream is
synchronized with the producer's stream before executing the consumer.

However in case of distributed autograd, the SendRpcBackward function receives
gradients over the wire and TensorPipe uses its own pool of streams for this
purpose. As a result, the tensors are received on TensorPipe's stream pool but
SendRpcBackward runs on a different stream during the backward pass and there
is no logic to synchronize these streams.

To fix this, I've enhanced DistEngine to synchronize these streams
appropriately when it receives grads over the wire.

Differential Revision: [D27025307](https://our.internmc.facebook.com/intern/diff/D27025307/)

### CODE REVIEW

File: torch/csrc/distributed/autograd/engine/dist_engine.cpp [hunk 1]
```diff
@@ -423,8 +424,27 @@ std::shared_ptr<c10::ivalue::Future> DistEngine::
...
+  // Typically the local autograd engine ensures stream synchronizations between
+  // nodes in the graph. However, for distributed autograd the sendFunction
+  // inputs might have been retrieved over the wire on a separate stream and the
+  // sendFunction itself runs on a different stream. As a result, we need to
+  // manually synchronize those two streams here.
+  const auto& send_backward_stream = sendFunction->stream(c10::DeviceType::CUDA);
+  if (send_backward_stream) {
+    for (const auto& grad : sendFunction->getGrads()) {
+        const auto guard = c10::impl::VirtualGuardImpl{c10::DeviceType::CUDA};
+        const auto default_stream = guard.getStream(grad.device());
```
Comment (rohan-varma) [inline torch/csrc/distributed/autograd/engine/dist_engine.cpp:439]: nit: guard.getStream() calls into `getCurrentCUDAStream` right, so this should actually be the current stream not necessarily default?

**Comment by pritamdamania87** [comment]
@ezyang @malfet This is a resubmit of https://github.com/pytorch/pytorch/pull/53769 since the mutigpu test failed on it. Although, I ran ci-all: https://github.com/pytorch/pytorch/pull/53941 and multigpu tests are passing for the same PR. I also logged into the circleci docker container for multigpu tests and ran the test 100 times and it passed.

I was wondering if there is something else I can do here, because I can't seem to reproduce the original issue which caused the PR to be reverted.

File: torch/csrc/distributed/autograd/engine/dist_engine.cpp (hunk 1 above)
Comment (pritamdamania87) [reply torch/csrc/distributed/autograd/engine/dist_engine.cpp:439]: Right, but we want the current stream and not the default stream. TensorPipe uses a bunch of custom streams to run the RPC function in, so we want to synchronize on the streams that TensorPipe used and not necessarily the default stream for the device.

**Comment by malfet** [comment]
@pritamdamania87 hmm, this PR fails multigpu test right now, isn't it? (See DrCI comment)

**Comment by pritamdamania87** [comment]
> @pritamdamania87 hmm, this PR fails multigpu test right now, isn't it? (See DrCI comment)

@malfet Right, but its a separate MPI issue which seems to be environment related rather than the test that this PR is introducing

**Comment by ezyang** [comment]
@pritamdamania This PR seems to no longer be failing multigpu, so I don't know if you did something. However, the failure @malfet pointed out seemed to point squarely at being induced by this PR (even if the bug was no fault of this PR) because the failure in https://www.internalfb.com/intern/opensource/ci/job/log/2958451454399960/ is *precisely* on the new test you added in this PR. Maybe you're hitting a latent bug in the MPI setup, but you're going to have to figure out what it is or skip this test on that configuration.

**Comment by pritamdamania87** [comment]
> @malfet pointed out seemed to point squarely at being induced by this PR (even if the bug was no fault of this PR) because the failure in https://www.internalfb.com/intern/opensource/ci/job/log/2958451454399960/ is precisely on the new test you added in this PR. Maybe you're hitting a latent bug in the MPI setup, but you're going to have to figure out what it is or skip this test on that configuration.

@ezyang I might be missing something here, but the test introduced in this PR is `test_remote_module`. Looking at the logs: https://www.internalfb.com/intern/opensource/ci/job/log/2958451454399960/, it doesn't seem like that test was even run. Isn't CI failing on `test_destroy_group` instead: 

```
Mar 12 23:19:52   test_destroy_group (__main__.TestMPIWithFork) ... [79b6db2c7988:609] *** An error occurred in MPI_Comm_create
Mar 12 23:19:52 [79b6db2c7988:609] *** reported by process [3889430529,0]
Mar 12 23:19:52 [79b6db2c7988:609] *** on communicator MPI_COMM_WORLD
Mar 12 23:19:52 [79b6db2c7988:609] *** MPI_ERR_TRUNCATE: message truncated
Mar 12 23:19:52 [79b6db2c7988:609] *** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
Mar 12 23:19:52 [79b6db2c7988:609] ***    and potentially your MPI job)
Mar 12 23:19:52 Traceback (most recent call last):
Mar 12 23:19:52   File "test/run_test.py", line 1075, in <module>
Mar 12 23:19:52     main()
Mar 12 23:19:52   File "test/run_test.py", line 1054, in main
Mar 12 23:19:52     raise RuntimeError(err_message)
Mar 12 23:19:52 RuntimeError: distributed/test_distributed_fork failed!
```

**Comment by ezyang** [comment]
You are right, I misread the log.

**Review by wanchaol — APPROVED** [review]
stamp for reland

**Comment by pritamdamania87** [comment]
@malfet I'm not sure what are the next steps in terms of landing this PR. As I mentioned in previous comments I did try to repro the original failures that caused the PR to break master, but couldn't and also it looks like ci-all is passing. Should we land the PR and just revert if it breaks master again?

**Pull request closed without merging**